src/librustc_codegen_ssa/back/write.rs

```diff
@@ -883,15 +821,12 @@ fn execute_copy_from_cache_work_item<B: ExtraBackendMethods>(
     }
 
     assert_eq!(object.is_some(), module_config.emit_obj != EmitObj::None);
-    assert_eq!(bytecode.is_some(), module_config.emit_bc);
-    assert_eq!(bytecode_compressed.is_some(), module_config.emit_bc_compressed);
 
     Ok(WorkItemResult::Compiled(CompiledModule {
         name: module.name,
         kind: ModuleKind::Regular,
         object,
-        bytecode,
-        bytecode_compressed,
+        bytecode: None,
```

Comment:
I'm not sure, I don't understand this part of the code well. Here's how I wrote this patch, in case it's helpful.

I started by deleting the `files.push((WorkProductFileKind::Bytecode, path.clone()));` line from `copy_all_cgu_workproducts_to_incr_comp_cache_dir`, like you suggested. That was the only place where a `WorkProductFileKind::Bytecode` value was produced, so I was able to remove that variant and its remaining uses.

That left just a single variant, `WorkProductFileKind::Object`, so I was able to also remove that variant and `WorkProductFileKind` completely.

The resulting code passes tests, but it is possible that there is some case not covered by tests. Can you expand on the case you are thinking of? Thanks.

Reply:
When using incremental mode combined with `--emit llvm-bc`, if the codegen unit is not cached, the bitcode will be written and the `CompiledModule` will get `bytecode` set to Some. However when the codegen unit is cached, the writing of the bitcode won't happen, as no separate bitcode file is cached. You would need to extract the bitcode from the object file and then write it to the right location.

```bash
$ rustc -Cincremental=incr --emit llvm-bc a.rs
$ ls
a.rs
a.bc
incr/
$ rm a.bc
$ rustc -Cincremental=incr --emit llvm-bc a.rs
a.rs
incr/
$ # a.bc is missing
```

Reply:
I tried running those commands but I got "warning: ignoring emit path because multiple .bc files were produced" and no `.bc` files were produced.

More generally, I clearly misunderstood the implications of your original suggestion. Perhaps I should just drop this commit from this PR, and it could be done as a follow-up.

Reply:
> I tried running those commands but I got "warning: ignoring emit path because multiple .bc files were produced" and no .bc files were produced.

It seems that incremental mode forces multiple codegen units, while `--emit llvm-bc` requires a single codegen unit. Using `--crate-type lib` fixes the error by only codegening a single module (the crate root) instead of two (the crate root and the `std::rt` module for `std::rt::lang_start`)

Reply:
I added `--crate-type lib` to the command sequence you gave above and it worked ok -- `a.bc` was re-generated by the second `rustc` command after having been deleted.